### PR TITLE
osmo-rtl-sdr: init at 2.0.1

### PR DESCRIPTION
**Pull request by skovati**
## Description of changes

`osmo-rtl-sdr` is a library for turning a RTL2832 based DVB dongle into a Software Defined Radio (SDR).

This propose package is the upstream of the existing `rtl-sdr` nixpkg, [which is a fork](https://github.com/librtlsdr/librtlsdr). This `rtl-sdr` fork no longer cuts releases, so the nixpkg is stuck on a 2020 release.

`osmo-rtl-sdr`, the new package this PR adds, instead directly builds the upstream project, which gets regularly versioned releases.

https://gitea.osmocom.org/sdr/rtl-sdr/
https://osmocom.org/projects/rtl-sdr/wiki

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/radio/osmo-rtl-sdr/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  cmakeFlags = lib.optionals stdenv.isLinux [
+    "-DINSTALL_UDEV_RULES=ON"
+    "-DWITH_RPC=ON"
+  ];
+
+  meta = with lib; {
+    description = "Software to turn the RTL2832U into a SDR receiver";
+    homepage = "https://gitea.osmocom.org/sdr/rtl-sdr";
+    license = licenses.gpl2Plus;
+    platforms = platforms.linux ++ platforms.darwin;
```
Comment (jcdickinson) [inline pkgs/applications/radio/osmo-rtl-sdr/default.nix:41]: maintainers meta is required, even if empty.

**Review by jcdickinson — CHANGES_REQUESTED** [review]
Update subject line of commit, and title of PR to:

```
osmo-rtl-sdr: init at 2.0.1
```

File: pkgs/applications/radio/osmo-rtl-sdr/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+, cmake
+, pkg-config
+, libusb1
+}:
+
+stdenv.mkDerivation rec {
+  pname = "osmo-rtl-sdr";
+  version = "2.0.1";
+
+  src = fetchgit {
```
Comment (MikaelFangel) [inline pkgs/applications/radio/osmo-rtl-sdr/default.nix:13]: Any reason your not using `fetchFromGitea`? (:

File: pkgs/applications/radio/osmo-rtl-sdr/default.nix (hunk 1 above)
Comment (MikaelFangel) [reply pkgs/applications/radio/osmo-rtl-sdr/default.nix:41]: And, mainProgram should also be specified if relevant. (Which it is in the most cases)

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -34533,6 +34533,8 @@ with pkgs;
 
   osmo = callPackage ../applications/office/osmo { };
 
+  osmo-rtl-sdr = callPackage ../applications/radio/osmo-rtl-sdr { };
```
Comment (MikaelFangel) [inline pkgs/top-level/all-packages.nix:34536]:
```suggestion
```

The package should be under the `pkgs/by-name/so/some-package/package.nix` hierarchy if possible. (This is the preferred way of adding new packages when possible)

**Comment by markuskowa** [comment]
We should probably consider to actually replace the old rtl-sdr package with this package (instead of adding as a new package). If I understand it correctly, the osmocom branch is a maintained branch of the same codebase.

**Comment by skovati** [comment]
Technically the currently packaged fork _is_ somewhat maintained, but they don't cut versioned releases anymore, which doesn't make it a clear cut answer as to which codebase should be _the_ package.

If we keep it as a separate package, I could also update the existing `rtl-sdr` fork to point to the latest commit on master, giving us up to date packages of both codebases.

If we overwrote `rtl-sdr` with the osmocom upstream, would we need to retest all the other SDR packages that depend on it?

@markuskowa

File: pkgs/applications/radio/osmo-rtl-sdr/default.nix (hunk 1 above)
Comment (skovati) [reply pkgs/applications/radio/osmo-rtl-sdr/default.nix]: Changes have been made

File: pkgs/applications/radio/osmo-rtl-sdr/default.nix (hunk 2 above)
Comment (skovati) [reply pkgs/applications/radio/osmo-rtl-sdr/default.nix]: Nope, made the switch!

File: pkgs/top-level/all-packages.nix (hunk 3 above)
Comment (skovati) [reply pkgs/top-level/all-packages.nix:34536]: Good call, I've moved this new package.

**Comment by markuskowa** [comment]
I think we should try first if we can simply upgrade the existing rtlsdr to the new osmocom sources. If that works and doesn't cause any breakage with packages using it, then there is no point in keeping two variants of the library around. 
The main upgrades in the library are probably the support of newer dongle versions. 

CC @bjornfor

**Comment by Tungsten842** [comment]
> I think we should try first if we can simply upgrade the existing rtlsdr to the new osmocom sources. If that works and doesn't cause any breakage with packages using it, then there is no point in keeping two variants of the library around. The main upgrades in the library are probably the support of newer dongle versions.
> 
> CC @bjornfor

The current fork has some improvements that are not present in the osmocom source: https://github.com/librtlsdr/librtlsdr/blob/master/README_improvements.md
I belive that the current source should be kept in nixpkgs as an alternative to the osmocom version.

**Comment by bjornfor** [comment]
Since I was pinged: I prefer to have only one version, if possible. (Assuming the old package is not maintained anymore.)

**Comment by skovati** [comment]
I'd agree with @Tungsten842 here. I think both codebases have merit and are separately maintained, and introducing a new package would guarantee no breaking changes. @bjornfor

**Comment by markuskowa** [comment]
> > I think we should try first if we can simply upgrade the existing rtlsdr to the new osmocom sources. If that works and doesn't cause any breakage with packages using it, then there is no point in keeping two variants of the library around. The main upgrades in the library are probably the support of newer dongle versions.
> > CC @bjornfor
> 
> The current fork has some improvements that are not present in the osmocom source: https://github.com/librtlsdr/librtlsdr/blob/master/README_improvements.md I belive that the current source should be kept in nixpkgs as an alternative to the osmocom version.

Ok, fair enough that may be an argument to have an alternative implementaion. However, from the cited URL it is difficult to judge what the improvements still are, since this document is already 3 years old.

File: pkgs/by-name/os/osmo-rtl-sdr/package.nix [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  cmakeFlags = lib.optionals stdenv.isLinux [
+    "-DINSTALL_UDEV_RULES=ON"
+    "-DWITH_RPC=ON"
+  ];
+
+  meta = with lib; {
+    description = "Software to turn the RTL2832U into a SDR receiver";
+    homepage = "https://gitea.osmocom.org/sdr/rtl-sdr";
+    license = licenses.gpl2Plus;
+    maintainers = [];
```
Comment (markuskowa) [inline pkgs/by-name/os/osmo-rtl-sdr/package.nix:43]: The package should have a maintainer. Do you want to add yourself here?

File: pkgs/by-name/os/osmo-rtl-sdr/package.nix [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    platforms = platforms.linux ++ platforms.darwin;
```
Comment (markuskowa) [inline pkgs/by-name/os/osmo-rtl-sdr/package.nix:44]:
```suggestion
    platforms = platforms.unix;
```
Better?

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -34533,6 +34533,8 @@ with pkgs;
 
   osmo = callPackage ../applications/office/osmo { };
 
+  osmo-rtl-sdr = callPackage ../by-name/os/osmo-rtl-sdr/package.nix { };
```
Comment (markuskowa) [inline pkgs/top-level/all-packages.nix:34536]: You still need to remove this line. The point of `pkgs/by-name` is to de-clutter `all-packages.nix`. 
See https://github.com/NixOS/rfcs/blob/master/rfcs/0140-simple-package-paths.md

**Review by markuskowa — CHANGES_REQUESTED** [review]

File: pkgs/by-name/os/osmo-rtl-sdr/package.nix [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+, pkg-config
+, libusb1
+}:
+
+stdenv.mkDerivation rec {
+  pname = "osmo-rtl-sdr";
+  version = "2.0.1";
+
+  src = fetchFromGitea {
+    domain = "https://gitea.osmocom.org";
```
Comment (markuskowa) [inline pkgs/by-name/os/osmo-rtl-sdr/package.nix:14]:
```suggestion
    domain = "gitea.osmocom.org";
```

File: pkgs/by-name/os/osmo-rtl-sdr/package.nix [hunk 8]
```diff
@@ -0,0 +1,47 @@
+{ lib
+, stdenv
+, fetchFromGitea
+, cmake
+, pkg-config
+, libusb1
+}:
+
+stdenv.mkDerivation rec {
+  pname = "osmo-rtl-sdr";
```
Comment (markuskowa) [inline pkgs/by-name/os/osmo-rtl-sdr/package.nix:10]: I think making the variant a suffix is maybe better?
```suggestion
  pname = "rtl-sdr-osmocom";
```

Please also move it then to the corresponding by-name path.

File: pkgs/by-name/os/osmo-rtl-sdr/package.nix [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+  propagatedBuildInputs = [ libusb1 ];
+
+  cmakeFlags = lib.optionals stdenv.isLinux [
+    "-DINSTALL_UDEV_RULES=ON"
+    "-DWITH_RPC=ON"
+  ];
+
+  meta = with lib; {
+    description = "Software to turn the RTL2832U into a SDR receiver";
```
Comment (markuskowa) [inline pkgs/by-name/os/osmo-rtl-sdr/package.nix:40]: If possible, you could use the `longDescription` field to give a very brief description what the difference to the "main rtl-sdr package" is (2 or three lines).

**Review by markuskowa — CHANGES_REQUESTED** [review]

**Comment by skovati** [comment]
Changes have been made @markuskowa. Thanks for all the great feedback!

**Comment by markuskowa** [comment]
@GrahamcOfBorg build rtl-sdr-osmocom

**Review by markuskowa — APPROVED** [review]

**Pull request merged**